test/unit/services/aws/credentials_service_test.rb

```diff
@@ -0,0 +1,76 @@
+# frozen_string_literal: true
+
+require 'test_helper'
+
+class Aws::CredentialsServiceTest < ActiveSupport::TestCase
+  setup do
+    File.stubs(:exist?).returns(true)
+  end
+
+  test '#call returns IAM credentials when available' do
+    Aws::AssumeRoleWebIdentityCredentials.expects(:new).never
+
+    assert Aws::CredentialsService.call(iam_auth_params), {
+      access_key_id: iam_auth_params[:access_key_id],
+      secret_access_key: iam_auth_params[:secret_access_key]
+    }
+  end
+
+  test '#call returns STS credentials when available' do
+    Aws::AssumeRoleWebIdentityCredentials.expects(:new).with(sts_auth_params).returns(assume_role_response)
+
+    assert Aws::CredentialsService.call(sts_auth_params), { credentials: assume_role_response }
+  end
+
+  test '#call uses a default role session name for STS credentials if not provided' do
+    Aws::AssumeRoleWebIdentityCredentials
+      .expects(:new)
+      .with(sts_auth_params.merge(role_session_name: '3scale-porta'))
+      .returns(assume_role_response)
+
+    assert Aws::CredentialsService.call(
+      sts_auth_params.except(:role_session_name)
+    ), { credentials: assume_role_response }
+  end
+
+  test '#call raises an error if the web_identity_token_file does not exist' do
+    File.stubs(:exist?).with(sts_auth_params[:web_identity_token_file]).returns(false)
+
+    Aws::AssumeRoleWebIdentityCredentials.expects(:new).never
+
+    assert_raises(Aws::CredentialsService::TokenNotFoundError) do
+      Aws::CredentialsService.call(sts_auth_params)
+    end
+  end
+
+  test '#call raises an error if not enough credential params are present' do
+    assert_raises(Aws::CredentialsService::AuthenticationTypeError) do
+      Aws::CredentialsService.call(full_params.except(:access_key_id, :role_arn))
+    end
+  end
+
+  private
+
+  def full_params
+    {
+      access_key_id: 'access_key_id',
+      secret_access_key: 'secret_access_key',
+      web_identity_token_file: '/path/to/token',
+      role_arn: 'role_arn',
+      role_session_name: 'role_session_name',
+      region: 'region'
+    }
+  end
+
+  def iam_auth_params
+    full_params.slice(:access_key_id, :secret_access_key)
+  end
+
+  def sts_auth_params
+    full_params.slice(:region, :role_arn, :role_session_name, :web_identity_token_file)
```

Comment:
I wonder, can't we do some call to get `region` instead of requiring a config value?

Reply:
Do you mean something like "calling Amazon to figure out the region"?
`region` is a mandatory value we send to S3 even with IAM authentication.

Reply:
Yes, IIRC there was a a call to check region of a bucket.

Reply:
I was looking into it and found [get-bucket-location](https://docs.aws.amazon.com/cli/latest/reference/s3api/get-bucket-location.html#get-bucket-location), which's part of the S3 API. It's also exposed [in the AWS Gem](https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/S3/Client.html#get_bucket_location-instance_method).
The catch here is that we need an [S3 Client](https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/S3/Client.html#initialize-instance_method) to perform this call, which depends on a `region` as well.
Now, before digging deeper on this one, is this really something we'd like to do? My reasoning is that the `region` is a required argument that needs to be provided in order for the application to communicate with S3, not that far from the actual credentials, in the sense that all values are mandatory to proceed.
Right now we request that the `region` is provided for IAM authentication and the same will still be true for STS auth, so I think it's not necessary to be changed... Also, even if this call is possible somehow, do we need to add yet another external call to the app?

Reply:
If it was a new requirement, I would insist. Given it has been like this before, we can keep it.

PS one more call, provided we will do a lot of calls, one more or less would not bother me personally.